mlflow/tracing/trace_manager.py

```diff
@@ -0,0 +1,182 @@
+        span_type: Optional[str] = None,
+    ) -> MLflowSpanWrapper:
+        """
+        Start a new OpenTelemetry span that is not part of the current trace context, but with the
+        explicit parent span ID if provided.
+
+        Args:
+            name: The name of the span.
+            trace_id: The trace ID for the span. Only used for getting the parent span for the given
+                parent_span_id. If not provided, a new trace will be created.
+            parent_span_id: The parent span ID of the span. If None, the span will be a root span.
+            span_type: The type of the span.
+
+        Returns:
+            The newly created span (wrapped in MLflowSpanWrapper). If any error occurs, returns a
+            NoOpMLflowSpanWrapper that has exact same interface but no-op implementations.
+        """
+        from mlflow.tracing.provider import get_tracer
+
+        try:
+            tracer = get_tracer(__name__)
+            if parent_span_id:
+                parent_span = self.get_span_from_id(trace_id, parent_span_id)._span
+                context = trace_api.set_span_in_context(parent_span)
+            else:
+                context = None
+
+            span = MLflowSpanWrapper(tracer.start_span(name, context=context), span_type=span_type)
+
+            self.add_or_update_span(span)
+            return span
+        except Exception as e:
+            _logger.warning(f"Failed to start span {name}: {e}")
+            return NoOpMLflowSpanWrapper()
+
+    def add_or_update_span(self, span: MLflowSpanWrapper):
+        """
+        Store the given span in the in-memory trace data. If the trace does not exist, create a new
+        trace with the trace_id of the span. If the span ID already exists in the trace, update the
+        span with the new data.
+
+        Args:
+            span: The span to be stored.
+        """
+        if not isinstance(span, MLflowSpanWrapper):
+            _logger.warning(f"Invalid span object {type(span)} is passed. Skipping.")
+            return
+
+        trace_id = span.trace_id
+        if trace_id not in self._traces:
+            # NB: the first span might not be a root span, so we can only
+            # set trace_id here. Other information will be propagated from
+            # the root span when it ends.
+            self._create_empty_trace(trace_id)
+
+        trace_data_dict = self._traces[trace_id].span_dict
+        trace_data_dict[span.span_id] = span
+
+    def _create_empty_trace(
+        self,
+        trace_id: str,
+        experiment_id: Optional[str] = None,
+        attributes: Optional[Dict[str, str]] = None,
+        tags: Optional[Dict[str, str]] = None,
+    ):
+        trace_info = TraceInfo(
+            trace_id=trace_id,
+            experiment_id=experiment_id
+            or "EXPERIMENT",  # TODO: Fetch this from global state or create a new one
+            start_time=None,
+            end_time=None,
```

Comment:
Can we make TraceInfo start_time and end_time optional and default to None?

Reply:
Great point, I think we shouldn't allow this when sending to backend, while it's convenient to allow in client side. Indeed, the [TraceInfo](https://github.com/mlflow/mlflow/blob/tracing/mlflow/entities/trace_info.py#L11) definition in `mlflow/entities` doesn't allow this. I'm going to file a PR for merging our dataclasses with those entities, so let me do the null handling in that PR🙂